src/Form/FormFieldContext.jsx

```diff
@@ -0,0 +1,93 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { newId } from '../utils';
+import useToggle from '../hooks/useToggle';
+import {
+  callAllHandlers,
+  useIdList,
+  mergeAttributeValues,
+  omitUndefinedProperties,
+} from './fieldUtils';
+import { FORM_CONTROL_SIZES } from './constants';
+
+const FormFieldContext = React.createContext();
+
+const defaultFormFieldContextValue = {
+  getNewDescriptorId: () => {}, // noop by default
+};
+
+const useFormFieldContext = (props = {}) => {
+  const cleanedProps = omitUndefinedProperties(props);
+  const contextValue = React.useContext(FormFieldContext) || {};
+  const mergedValues = {
+    ...defaultFormFieldContextValue,
+    ...contextValue,
+    ...cleanedProps,
+    'aria-describedby': mergeAttributeValues(
+      contextValue['aria-describedby'],
+      props['aria-describedby'],
+    ),
+    onFocus: callAllHandlers(contextValue.onFocus, props.onFocus),
+    onBlur: callAllHandlers(contextValue.onBlur, props.onBlur),
+  };
+  return mergedValues;
+};
+
+const FormFieldContextProvider = ({
+  children,
+  id,
+  isInvalid,
+  isValid,
+  onFocus,
+  onBlur,
+  size,
+}) => {
+  const resolvedId = React.useMemo(() => id || newId('form-field'), [id]);
```

Comment:
Should we also use `newId` with a user-provided id, or are we just trusting people to not mess it up?

Reply:
I think trusting is better. I believe newId can cause issues for some server rendered react apps (gatsby) since the index gets reset to zero when the page is loaded. I'm not sure what prospectus has been doing about this, but this would be their escape hatch, if it is indeed and issue.